apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/SystemMetrics.java

```diff
@@ -78,26 +97,133 @@
+                        lineCgroup = cgroupLine;
+                    }
+                    if (MEMORY_CGROUP.matcher(cgroupLine).matches()) {
+                        lineCgroup = cgroupLine;
+                    }
+                }
+                if (lineCgroup != null) {
+                    CgroupFiles cgroupFilesTest = null;
+                    try(BufferedReader fileMountInfoReader = new BufferedReader(new FileReader(mountInfo))) {
+                        for (String mountLine = fileMountInfoReader.readLine(); mountLine != null && !mountLine.isEmpty(); mountLine = fileMountInfoReader.readLine()) {
+                            String foundRegex = applyCgroup2Regex(mountLine);
+                            if (foundRegex != null) {
+                                cgroupFilesTest = verifyCgroup2Available(lineCgroup, new File(foundRegex));
+                                if (cgroupFilesTest != null) return cgroupFilesTest;
+                            }
+                            foundRegex = applyCgroup1Regex(mountLine);
+                            if (foundRegex != null) {
+                                cgroupFilesTest = verifyCgroup1Available(new File(foundRegex));
+                                if (cgroupFilesTest != null) return cgroupFilesTest;
+                            }
+                        }
+                    }
+                    // Fall back to /sys/fs/cgroup if not found on mountinfo
+                    cgroupFilesTest = verifyCgroup2Available(lineCgroup, new File(SYS_FS_CGROUP));
+                    if (cgroupFilesTest != null) return cgroupFilesTest;
+                    cgroupFilesTest = verifyCgroup1Available( new File(SYS_FS_CGROUP + File.pathSeparator + "memory"));
+                    if (cgroupFilesTest != null) return cgroupFilesTest;
+
+                }
+            } catch (Exception e) {
+            }
+        }
+        return null;
+    }
+    String applyCgroup1Regex(String mountLine) {
+        Matcher matcher = CGROUP1_MOUNT_POINT.matcher(mountLine);
+        if (matcher.matches()) {
+            return matcher.group(1);
+        }
+        return null;
+    }
+    String applyCgroup2Regex(String mountLine) {
+        Matcher matcher = CGROUP2_MOUNT_POINT.matcher(mountLine);
+        if (matcher.matches()) {
+            return matcher.group(1);
+        }
+        return null;
+    }
+    private CgroupFiles verifyCgroup2Available(String lineCgroup, File mountDiscovered) throws IOException {
+        final String[] cgroupSplit = StringUtils.split(lineCgroup, ':');
+        // Checking cgroup2
+        File maxMemory = new File(mountDiscovered, cgroupSplit[cgroupSplit.length - 1] + "/" + CGROUP2_MAX_MEMORY);
+        if (maxMemory.canRead()) {
+            try(BufferedReader fileReaderMem = new BufferedReader(new FileReader(maxMemory))) {
+                String memMaxLine = fileReaderMem.readLine();
+                if (!"max".equalsIgnoreCase(memMaxLine)) {
```

Comment:
Ohh, so in cgroup2 `max` represents unlimited memory for the cgroup?
This is important! We just decided to transfer the special treatment of unlimited max to the UI, but we rely on sending the special numeric representing unlimited. We can't send a string. We can send the unlimited value, but it makes more sense, we better not send anything if we see `max` and make sure UI properly deals with that.

Reply:
Yes, on cgroup2 it is a string. what about to transfer the UNLIMITED constant from cgroup1 ?

Reply:
Yes, that's what I poorly tried to say in my comment - we could do that, but it doesn't make a lot of sense...
Letting the UI even be aware of the UNLIMITED numeric value was a way to implement once, instead of making all APM agents aware of that. Now that I am aware of the literal UNLIMITED value, we must handle in agents, as the metric value in the intake API is numeric.
I think that making sure agents omit the value when it is `max` and making sure UI expects that is enough.